docs: document trusted boot (#2612)
spectrocloud/librarium: docs/docs-content/clusters/edge/trusted-boot/deployment-day2/install.md

  requires you to have a wired keyboard and a monitor to display the interface connected to the Edge device.
- A USB disk containing the installer ISO.

:::warning

Before proceeding with installation, check the maximum EFI size that your Edge host is able to boot and compare it
against the installer ISO and provider image. With secure boot enabled and keys enrolled, you won't be able to boot from
the EFI size checking ISO unless you disable secure boot or clear the keys. For more information, refer to
[Check EFI Size and Edge Host Boot Limit](../edgeforge/check-efi-limit.md).

:::

## Instructions

### Enroll Secure Boot Keys into Edge Device

The following steps are written for the Intel NUC 13 Pro device. If you are using a different device, the exact sequence
of steps might be different. However, you should be able to find the same or similarly worded options in any device that
supports Secure Boot.

1. Insert the USB disk into the Edge device.

2. Power up your Edge device and enter the BIOS interface. Oftentimes you can accomplish this by pressing F2, F1, or F10

>    on the keyboard immediately after powering up the deice and before you select a boot volume, as is the case in Intel

Reviewer: ```suggestion
   on the keyboard immediately after powering up the device and before you select a boot volume, as is the case in Intel
```

   NUC 13 Pro. However, the exact method might be different on your Edge device. Consult the manufacturer of your Edge
   device to find out how to enter the BIOS interface.

3. From the BIOS menu, select the **Boot** tab. Then select **Secure Boot**.

4. From the **Secure Boot Menu**, make sure that secure boot is enabled. Then select **Restore to Setup Mode**. Two
   dialogue boxes will pop up to confirm this operation. Select **Yes** to confirm. This will make the device restart.

   :::info

   **User Mode** and **Setup Mode** are UEFI terms. Setting secure boot mode to **Setup Mode** means clearing the keys
   that are currently in the device. This allows you to enroll your own keys, while **User Mode** means you have
   enrolled your own keys.

>    In some BIOS, the same actions or functionality might be named differently in the BIOS interface. Consult the

Reviewer: BIOSs?  or instead say "On some edge hosts, ..."

   manufacturer of your Edge host to find out how to clear the keys and enroll your own keys.

   :::

If you set the `AUTO_ENROLL_SECUREBOOT_KEYS` argument to `true` when you built the installation ISO, key enrollment will

> begin automatically once the device restarts. You can skip the following steps to proceed to

Reviewer: ```suggestion
begin automatically the first time the device boots from the installer ISO. You can skip the following steps to proceed to
```

Reviewer 2: Note: this option will be disregarded in VM. So in VMs, keys will always be auto enrolled.

[Install Palette Edge with Trusted Boot](#install-palette-edge-with-trusted-boot).

5. If you did not set `AUTO_ENROLL_SECUREBOOT_KEYS` to `true` when you built the ISO, you will need to select the
   **Enroll Secure Boot keys: Auto** option in the boot menu when the Edge device restarts. This will start the key
   enrollment process.

### Install Palette Edge with Trusted Boot

6. When the keys are finished enrolling, the installation will start automatically after a reboot. Installation
   typically takes about 5 minutes, but can vary depending on your hardware.

7. When installation finishes, you will observe the text "Installation has finished, rebooting in 5 seconds" on your

>    screen. Remove the USB disk from your Edge device.

Reviewer: at this point, from a security perspective, it is also advisable to ensure the boot order places the target installation drive first.

Author: This would require that the users go into the BIOS menu again, right? Can you elaborate a bit why this is necessary?

Reviewer: because you don't want the box to ever boot from anything except the main system drive post installation. attackers will try to oinsert usb disks and see if you did for some reason left usb as the first boot option. 
this is an important part of preparing your hardware for production. all other boot options should be disabled or removed from teh boot order once installation is completed.

Reviewer: for a better and more secure flow, what I do is this:
- I leave the system drive as the first boot option
- I manually in the bios tell it to boot off the usb stick during enrollment/installation (yes, it means I need to do that a couple of times)
- I look to see if I can remove any other potential boot options such as network boot, usb, etc.

We don't need to go into details here, but we should give the high level message to people that after installation, they should ensure that their device boot options are set securely and only the system drive should be bootable.

Reviewer: Note - this does not say that trusted boot wouldn't prevent booting a rogue usb stick, this is called defence in depth - just because we have trusted boot doesn't mean we (or our customers) shouldn't take all normal security precautions during device setup.

8. After installation finishes, power up the device. Let your device automatically choose the boot volume. It should
   boot directly to Edge device registration.

## Validate

Use the following steps to validate that Trusted Boot has been enabled on your Edge host. This requires you to have
configured a user on your Edge host. Refer to [Installer Reference](../../edge-configuration/installer-reference.md) for
more information.

1. Press **Alt + right arrow key**, or **Ctrl + Alt + F1**. Replace **Alt** with **Options** on a Mac keyboard. This
   will bring up a terminal and allow you to log in with the credentials you configured with the **user-data** file.

   Alternatively, you can establish an SSH connection to your Edge device if you have network access to it.

2. Issue the following command.

   ```shell

>    ls -ltr / run/cos/

Reviewer: should it be /run/cos ? there is space after the first /

---
sidebar_label: "Installation"
title: "Installation with Trusted Boot"
description: "Learn about how to install Palette Edge with Trusted Boot on your Edge device."
hide_table_of_contents: false
sidebar_position: 50
tags: ["edge"]
---

This page guides you through installing Palette Edge with Trusted Boot enabled on an Edge device. Some of the steps
included in this guide involve the Basic Input/Output System (BIOS) interface of the Edge device which varies from
machine to machine. This guide uses the Intel NUC 13 Pro as an example.

## Prerequisites

- Edge Installer ISO file. Check out the [Build Edge Artifacts with Trusted Boot](../edgeforge/build-trusted-iso.md) to
  learn how to create an Edge Installer ISO with Trusted Boot.
- The ability to modify the boot order settings to boot from a USB drive and to enter the BIOS interface. Typically this
  requires you to have a wired keyboard and a monitor to display the interface connected to the Edge device.
- A USB disk containing the installer ISO.

> :::warning

Reviewer 3: Excellent warning and crosslinking

Before proceeding with installation, check the maximum EFI size that your Edge host is able to boot and compare it
against the installer ISO and provider image. With secure boot enabled and keys enrolled, you won't be able to boot from
the EFI size checking ISO unless you disable secure boot or clear the keys. For more information, refer to
[Check EFI Size and Edge Host Boot Limit](../edgeforge/check-efi-limit.md).

:::

## Instructions

> ### Enroll Secure Boot Keys into Edge Device

Reviewer 3: Back to back headings feels odd. Can we add some filler text or potentially remove one subheading?

   :::info

   **User Mode** and **Setup Mode** are UEFI terms. Setting secure boot mode to **Setup Mode** means clearing the keys
   that are currently in the device. This allows you to enroll your own keys, while **User Mode** means you have
   enrolled your own keys.

   On some Edge hosts, the same actions or functionality might be named differently in the BIOS interface. Consult the
   manufacturer of your Edge host to find out how to clear the keys and enroll your own keys.

   :::

   If you set the `AUTO_ENROLL_SECUREBOOT_KEYS` argument to `true` when you built the installation ISO, key enrollment
   will begin automatically the first time the device boots from the installer ISO. You can skip the following steps to
   proceed to [Install Palette Edge with Trusted Boot](#install-palette-edge-with-trusted-boot). In Virtual Machines
   (VM), keys will always be auto-enrolled.

5. If you did not set `AUTO_ENROLL_SECUREBOOT_KEYS` to `true` when you built the ISO, you will need to select the
   **Enroll Secure Boot keys: Auto** option in the GRUB menu when the Edge host restarts. This will start the key
   enrollment process.

   :::warning

>    There is currently a known issue where if `AUTO_ENROLL_SECUREBOOT_KEYS` is set to `false`, it's possible for manual

Reviewer 3: We need to capture this in the global known issues page and add it to the release notes.

   key enrollment to take a few attempts to be successful.

   :::

### Install Palette Edge with Trusted Boot

6. When the keys are finished enrolling, the installation will start automatically after a reboot. Installation
   typically takes about 5 minutes, but can vary depending on your hardware.

> 7. When installation finishes, you will observe the text "Installation has finished, rebooting in 5 seconds" on your
>    screen. Remove the USB disk from your Edge device.

Reviewer 3: ```suggestion
7. When the installation finishes, the terminal will display the text "Installation has finished, rebooting in 5 seconds." Remove the USB disk from your Edge device.
```
Changed the phrasing to account for accessibility use cases.

   :::info

   If you have more than one boot devices, it is good security practice to go into the BIOS again and ensure that the
   target installation disk is prioritized in the boot sequence.

   :::

> 8. After installation finishes, power up the device. Let your device automatically choose the option from the GRand

Reviewer 3: lol I though GRand was a typo but that is actually how it is spelled out

>    Unified Bootloader (GRUB) menu. It should boot directly to Edge device registration.

Reviewer 3: It might be good to link to the registration page or at least mention something about that process. Only because in the Next Steps, it's not mentioned at all, so I want to avoid a user jumping straight to cluster creation without the registration process. Not everyone will use the auto-registration method.

configured a user on your Edge host. Refer to [Installer Reference](../../edge-configuration/installer-reference.md) for
more information.

1. Press **Alt + right arrow key**, or **Ctrl + Alt + F1**. Replace **Alt** with **Options** on a Mac keyboard. This
   will bring up a terminal and allow you to log in with the credentials you configured with the **user-data** file.

   Alternatively, you can establish an SSH connection to your Edge device if you have network access to it.

2. Issue the following command.

   ```shell
   ls -ltr /run/cos/
   ```

   You should receive an output that's similar to the following.

   ```
   total 12
   -rwxr-xr-x  1 root root 1 Apr 30 15:59 uki_boot_mode
   -rwxr-xr-x  1 root root  1 Apr 30 15:59 active_mode
   -rw-r--r--  1 root root 534 Apr 30 15:59 cos-layout.env
   ```

   This confirms that your Edge device was booted with a Unified Kernel Image, a specific file format tailored to
   achieve a tamper-proof system and encryption of persistent partitions.

3. Issue the following command to verify disk encryption.

   ```shell
   lsblk
   ```

>    You will observe output similar to the following and verify that the **/oem** portion and various directories
>    containing sensitive user data are encrypted. This means that if the boot process is ever altered by an unauthorized
>    party, they will not be able to gain access to the encrypted data.

Reviewer 3: ```suggestion
   An output similar to the following is displayed. Verify that the **/oem** portion and various directories
   containing sensitive user data are encrypted. This means that if the boot process is ever altered by an unauthorized
   party, they will not be able to gain access to the encrypted data.
```

@lennessyy looking the output, how can you determine it's encrypted? That was not intuitive to me.
